Modules/_queuemodule.c

```diff
@@ -0,0 +1,422 @@
+        microseconds = -1;
+    }
+
+    /* put() signals the queue to be non-empty by releasing the lock.
+     * So we simply try to acquire the lock in a loop, until the condition
+     * (queue non-empty) becomes true.
+     */
+    while (self->lst_pos == PyList_GET_SIZE(self->lst)) {
+        /* First a simple non-blocking try without releasing the GIL */
+        r = PyThread_acquire_lock_timed(self->lock, 0, 0);
+        if (r == PY_LOCK_FAILURE && microseconds != 0) {
+            Py_BEGIN_ALLOW_THREADS
+            r = PyThread_acquire_lock_timed(self->lock, microseconds, 1);
+            Py_END_ALLOW_THREADS
+        }
+        if (r == PY_LOCK_INTR && Py_MakePendingCalls() < 0) {
+            return NULL;
+        }
+        if (r == PY_LOCK_FAILURE) {
+            /* Timed out */
+            PyErr_SetNone(EmptyError);
+            return NULL;
+        }
+        self->locked = 1;
+        /* Adjust timeout for next iteration (if any) */
+        if (endtime > 0) {
+            timeout_val = endtime - _PyTime_GetMonotonicClock();
+            microseconds = _PyTime_AsMicroseconds(timeout_val, _PyTime_ROUND_CEILING);
+        }
+    }
+    /* BEGIN GIL-protected critical section */
+    assert(self->lst_pos < PyList_GET_SIZE(self->lst));
+    item = simplequeue_pop_item(self);
+    if (self->locked) {
+        PyThread_release_lock(self->lock);
+        self->locked = 0;
+    }
+    /* END GIL-protected critical section */
+
+    return item;
+}
+
+/*[clinic input]
+_queue.SimpleQueue.get_nowait
+
+Remove and return an item from the queue without blocking.
+
+Only get an item if one is immediately available. Otherwise
+raise the Empty exception.
+[clinic start generated code]*/
+
+static PyObject *
+_queue_SimpleQueue_get_nowait_impl(simplequeueobject *self)
+/*[clinic end generated code: output=a89731a75dbe4937 input=6fe5102db540a1b9]*/
+{
+    return _queue_SimpleQueue_get_impl(self, 0, Py_None);
+}
+
+/*[clinic input]
+_queue.SimpleQueue.empty
+
+Return True if the queue is empty, False otherwise (not reliable!).
+[clinic start generated code]*/
+
+static PyObject *
+_queue_SimpleQueue_empty_impl(simplequeueobject *self)
+/*[clinic end generated code: output=25164355586aaad0 input=990412cc070aa6a2]*/
+{
+    PyObject *r;
+    if (self->lst_pos == PyList_GET_SIZE(self->lst))
+        r = Py_True;
+    else
+        r = Py_False;
+    Py_INCREF(r);
+    return r;
+}
+
+/*[clinic input]
+_queue.SimpleQueue.qsize
+
+Return the approximate size of the queue (not reliable!).
+[clinic start generated code]*/
+
+static PyObject *
+_queue_SimpleQueue_qsize_impl(simplequeueobject *self)
+/*[clinic end generated code: output=4c6f52e316a53f60 input=76444ac3249e90c5]*/
+{
+    return PyLong_FromSsize_t(PyList_GET_SIZE(self->lst) - self->lst_pos);
+}
+
+static PyObject *
+simplequeue_sizeof(simplequeueobject *self, void *unused)
+{
+    Py_ssize_t res;
+
+    res = _PyObject_SIZE(Py_TYPE(self));
+    res += _PySys_GetSizeOf(self->lst);
```

Comment:
Since the list is visited in tp_traverse, its size shouldn't be include in the result of `__sizeof__`.

Reply:
Hmm... what does tp_traverse have to do with it?

Reply:
`sys.getsizeof()` is a low-level tool that returns the only size of the object, without subobjects. A high-level tool will calls `sys.getsizeof()` recursively for the tree of objects using `gc.get_referents()`. `gc.get_referents()` is build by using tp_traverse. If `__sizeof__` includes the size of objects exposed in tp_traverse, it will be counted twice.

Reply:
That's a fair point, even though any non-trivial `__sizeof__` defined in pure Python would have the same problem (see for example the pure-Python `OrderedDict.__sizeof__`).

I'll make the change.

Reply:
`OrderedDict.__sizeof__` is incorrect. In general pure Python classes shouldn't define `__sizeof__` (as well as `__basicsize__` or `__dictoffset__`).